### PR TITLE
fix protected event handler

### PR DESCRIPTION
**Pull request by TyOverby**
Fixes https://github.com/dotnet/roslyn/issues/20335

### CODE REVIEW

**Comment by TyOverby** [comment]
cc @dotnet/roslyn-compiler

**Comment by TyOverby** [comment]
Ping @dotnet/roslyn-compiler

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb [hunk 1]
```diff
@@ -24,6 +24,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
     Public Class EventSymbolTests
         Inherits BasicTestBase
 
+        <WorkItem(20335, "https://github.com/dotnet/roslyn/issues/20335")>
+        <Fact()>
+        Public Sub ProtectedHandler()
+            Dim source = <compilation name="F">
+                             <file name="F.vb">
+ Public Class Form1
+    Protected Event EnterOriginal As System.Action
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb]: Do we have a test that `MyBase` works for other accessibility modifiers: Friend for example? If not we should add them here.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 2]
```diff
@@ -2279,12 +2279,15 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                         hasErrors = True
                     Else
                         Dim useSiteDiagnostics As HashSet(Of DiagnosticInfo) = Nothing
+                        Dim accessThroughType = GetAccessThroughType(actualEventAccess.ReceiverOpt)
+                        If Not (Me.IsAccessible(method, useSiteDiagnostics, accessThroughType) OrElse
+                            Not IsAccessible(eventSymbol, useSiteDiagnostics, accessThroughType)) Then
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Why are we checking both the `method` and `eventSymbol` here for accessibility? I don't believe that VB events can declare custom accessibility hence I would assume that we should be checking only one here.

@cston ?

Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Might be easier to read as `If Not IsAccessible(method, ...) AndAlso IsAccessible(eventSymbol, ...) Then`

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb [hunk 3]
```diff
@@ -24,6 +24,32 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
...
+
+    Protected Property P1 As Integer
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb]: Please add a test that uses custom events in VB as well to ensure this works.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb (hunk 2 above)
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Agreed, we should only need the first check: `IsAccessible(method, useSiteDiagnostics, accessThroughType)`.

Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Would it not be an issue if they are referencing events from outside of VB?

Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: > Would it not be an issue if they are referencing events from outside of VB?

Possibly. If that is a scenario that matters though why didn't you add tests for it?

Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: > If that is a scenario that matters though why didn't you add tests for it?

Didn't think about it until this conversation chain brought it up :P

Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: ha! Glad we brought it up then. I hadn't thought about it either.

Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Yay!  Code reviews working as intended!

File: src/Compilers/VisualBasic/Test/Symbol/BasicCompilerSymbolTest.vbproj [hunk 4]
```diff
@@ -39,6 +39,10 @@
       <Project>{7FE6B002-89D8-4298-9B1B-0B5C247DD1FD}</Project>
       <Name>CompilerTestResources</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\..\Test\Utilities\CSharp\CSharpCompilerTestUtilities.csproj">
+      <Project>{4371944a-d3ba-4b5b-8285-82e5ffc6d1f9}</Project>
+      <Name>CSharpCompilerTestUtilities</Name>
+    </ProjectReference>
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Symbol/BasicCompilerSymbolTest.vbproj]: Is this reference necessary? There is a `CommonTestBase.CreateCSharpCompilation` method that should be available already.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb [hunk 5]
```diff
@@ -2279,12 +2279,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
                         Dim useSiteDiagnostics As HashSet(Of DiagnosticInfo) = Nothing
-
-                        If Not Me.IsAccessible(method, useSiteDiagnostics, If(actualEventAccess.ReceiverOpt IsNot Nothing, actualEventAccess.ReceiverOpt.Type, eventSymbol.ContainingType)) Then
+                        Dim accessThroughType = GetAccessThroughType(actualEventAccess.ReceiverOpt)
+                        If Not Me.IsAccessible(method, useSiteDiagnostics, accessThroughType) Then
                             Debug.Assert(eventSymbol.DeclaringCompilation IsNot Me.Compilation)
                             ReportDiagnostic(diagnostics, node.EventExpression, GetInaccessibleErrorInfo(method))
                         End If
 
+
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Extra blank line.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb [hunk 6]
```diff
@@ -23,6 +23,87 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests
...
+        RemoveHandler MyBase.EventB, Nothing
+                      ~~~~~~~~~~~~~
+BC30389: 'Form1.EventB' is not accessible in this context because it is 'Private'.
+        AddHandler EventB, Nothing
+                   ~~~~~~
+BC30389: 'Form1.EventB' is not accessible in this context because it is 'Private'.
+        RemoveHandler EventB, Nothing
+                      ~~~~~~
+</Expected>)
+        End Sub
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb]: Consider testing accessors with `Friend` accessibility and with different accessibility for event and accessor. For instance:
```
// Assembly A
.class public auto ansi beforefieldinit A
{
  .method assembly hidebysig newslot specialname virtual instance void 
          add_E(class [mscorlib]System.Action`1<int32> 'value') cil managed
  {
    ret
  }
  .method public hidebysig newslot specialname virtual instance void 
          remove_E(class [mscorlib]System.Action`1<int32> 'value') cil managed
  {
    ret
  }
  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    ldarg.0
    call       instance void [mscorlib]System.Object::.ctor()
    ret
  }
  .event class [mscorlib]System.Action`1<int32> E
  {
    .addon instance void A::add_E(class [mscorlib]System.Action`1<int32>)
    .removeon instance void A::remove_E(class [mscorlib]System.Action`1<int32>)
  }
}
// Assembly B
Class B
    Inherits A
    Sub M()
        AddHandler MyBase.E, Nothing 
        RemoveHandler MyBase.E, Nothing 
    End Sub 
End Class
```

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb (hunk 1 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb]: Done.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb (hunk 3 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb]: Done

File: src/Compilers/VisualBasic/Test/Symbol/BasicCompilerSymbolTest.vbproj (hunk 4 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Symbol/BasicCompilerSymbolTest.vbproj]: Oh neat, thanks!

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb (hunk 5 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Statements.vb]: Fixed

File: src/Compilers/VisualBasic/Test/Symbol/BasicCompilerSymbolTest.vbproj (hunk 4 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Symbol/BasicCompilerSymbolTest.vbproj]: Fixed

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb (hunk 6 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Source/EventTests.vb]: Done.

**Comment by cston** [comment]
LGTM

**Comment by TyOverby** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by TyOverby** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by TyOverby** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by jaredpar** [comment]
retest windows_release_vs-integration_prtest please

**Pull request merged**